src/lib.rs

```diff
@@ -55,32 +58,64 @@ use quote::quote;
 ///     })().map_err(|err| err.context("context"))
 /// }
 /// ```
+///
+/// Sometimes you will receive borrowck errors, especially when returning references. These can
+/// often be fixed by setting the `move` option of the attribute macro. For example:
+///
+/// ```
+/// use fn_error_context::context;
+///
+/// #[context(move, "context")]
+/// fn returns_reference(val: &mut u32) -> anyhow::Result<&mut u32> {
+///     Ok(&mut *val)
+/// }
+/// ```
 #[proc_macro_attribute]
 pub fn context(args: TokenStream, input: TokenStream) -> TokenStream {
-    let args: proc_macro2::TokenStream = args.into();
+    let Args(move_token, format_args) = syn::parse_macro_input!(args);
     let mut input = syn::parse_macro_input!(input as syn::ItemFn);
 
     let body = &input.block;
     let return_ty = &input.sig.output;
-    if input.sig.asyncness.is_some() {
-        match return_ty {
+    let new_body = if input.sig.asyncness.is_some() {
+        let return_ty = match return_ty {
             syn::ReturnType::Default => {
-                return syn::Error::new_spanned(return_ty, "function should return Result")
+                return syn::Error::new_spanned(input, "function should return Result")
                     .to_compile_error()
-                    .into()
-            }
-            syn::ReturnType::Type(_, return_ty) => {
-                input.block.stmts = syn::parse_quote!(
-                    let result: #return_ty = async { #body }.await;
-                    result.map_err(|err| err.context(format!(#args)).into())
-                );
+                    .into();
             }
+            syn::ReturnType::Type(_, return_ty) => return_ty,
+        };
+        quote! {
+            let result: #return_ty = async #move_token { #body }.await;
+            result.map_err(|err| err.context(format!(#format_args)).into())
         }
     } else {
-        input.block.stmts = syn::parse_quote!(
-            (|| #return_ty #body)().map_err(|err| err.context(format!(#args)).into())
-        );
-    }
+        quote! {
+            // Moving a non-`Copy` value into the closure tells borrowck to always treat the closure
+            // as a `FnOnce`, preventing some borrowing errors.
+            let force_fn_once = ::core::iter::empty::<()>();
```

Comment:
I think we need to use an ugly name like `__force_fn_once` or it might conflict with variable in the function body

I did a quick search and it sounds like we could instead do `Ident::new("force_fn_once", Span::mixed_site())` (https://github.com/rust-lang/rust/pull/64690) which would be a bit cleaner

Reply:
I don't think this is _technically_ necessary because the variable is moved out of before any user code is run, but I suppose it's good practice to be hygienic :)